### PR TITLE
GOD-471 update to connector template

### PR DESCRIPTION
**Pull request by IharKazakou**

### CODE REVIEW

**Review by andrewi-wd — COMMENTED** [review]
Cannot start PREnv. Looking into it currently.

```
today at 10:23:03 AMContainer started
today at 10:22:02 AM2023-04-18T07:22:02: PM2 log: PM2 successfully stopped
today at 10:23:03 AM2023-04-18T07:23:03: PM2 log: Launching in no daemon mode
today at 10:23:03 AM2023-04-18T07:23:03: PM2 log: App [pm2-metrics:0] starting in -fork mode-
today at 10:23:03 AM2023-04-18T07:23:03: PM2 log: App [pm2-metrics:0] online
today at 10:23:03 AM2023-04-18T07:23:03: PM2 error: Script not found: /srv/server.js
today at 10:23:03 AM2023-04-18T07:23:03: PM2 log: Stopping app:pm2-metrics id:0
today at 10:23:03 AM2023-04-18T07:23:03: PM2 log: App [pm2-metrics:0] exited with code [0] via signal [SIGINT]
today at 10:23:04 AM2023-04-18T07:23:04: PM2 log: pid=19 msg=process killed
today at 10:23:04 AM2023-04-18T07:23:04: PM2 log: PM2 successfully stopped
today at 10:23:04 AMContainer stopped
```

File: src/integrations/ditto/client/APIAbstract.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import { Client } from 'undici'
+import type { Dispatcher, errors } from 'undici'
+
+import { AuthFailedError } from '../../../infrastructure/errors/publicErrors'
+
+import type { ApiRequest } from './types'
+import { isErrorResponse } from './types'
+
+export class APIAbstract {
```
[the remark this answers is not in the document]
Comment (IharKazakou) [reply src/integrations/ditto/client/APIAbstract.ts]: I am using node-core. Just forgot to delete this class. fixed

File: src/integrations/ditto/client/APIDitto.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import type { Dependencies } from '../../../infrastructure/diConfig'
+import { AuthFailedError, UnrecognizedError } from '../../../infrastructure/errors/publicErrors'
+
+import type {
+  UpdateVariantRequestBody,
+  UpdateVariantResult,
+  WorkspaceComponentsByName,
+} from './types'
+
+export class APIDitto {
```
[the remark this answers is not in the document]
Comment (IharKazakou) [reply src/integrations/ditto/client/APIDitto.ts:13]: What do you mean?  As I understand node-core handles that. Am I right?

[the remark this answers is not in the document]
Comment (IharKazakou) [reply src/integrations/ditto/client/APIDitto.ts:13]: ah I see that I should pass retry config

[the remark this answers is not in the document]
Comment (kibertoad) [reply src/integrations/ditto/client/APIDitto.ts:13]: yeah, default config is no retries

File: src/app.ts [hunk 3]
```diff
@@ -1,84 +1,143 @@
...
+
+export type ConfigOverrides = {
+  enableMetrics?: boolean
+}
+
+export async function getApp(configOverrides: ConfigOverrides = {}) {
+  const config = getConfig()
+  const appConfig = config.app
+  const loggerConfig = resolveLoggerConfiguration(appConfig)
+  const enableRequestLogging = ['debug', 'trace'].includes(appConfig.logLevel)
```
Comment (Serzhs) [inline src/app.ts:45]: 🟢 I suggest we add `['debug', 'trace']` in a constant

**Review by Serzhs — APPROVED** [review]
Generally speaking, more newlines and variables should be needed to improve readability. Walking through the code, my eyes didn't catch any red flags, so from my point of view. We can proceed :)

File: src/integrations/ditto/client/APIDitto.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (IharKazakou) [reply src/integrations/ditto/client/APIDitto.ts:13]: fixed

File: Dockerfile [hunk 4]
```diff
@@ -1,9 +1,9 @@
-FROM node:18.8.0-alpine3.16 as base
```
Comment (m-evans-1) [inline Dockerfile:1]: 🟠 This Dockerfile in root directory is obsolete, but the one in `docker/` is not up to date

File: Dockerfile-dev [hunk 5]
```diff
@@ -1,4 +1,4 @@
-FROM node:18-alpine
+FROM node:18.15.0-alpine3.17
```
Comment (m-evans-1) [inline Dockerfile-dev:1]: 🟠 `docker-compose.yml` in root should be updated with:


``` 
   ports:
      - '${HOST_APP_PORT}:3000'
      - '${HOST_PRIVATE_APP_PORT}:9080'

File: .env.example [hunk 6]
```diff
@@ -2,8 +2,23 @@ APP_NAME=app
...
 
+# app
+APP_VERSION=1
+GIT_COMMIT_SHA=sha
 APP_PORT=3000
-APP_ENV=production
+APP_BIND_ADDRESS=0.0.0.0
+APP_ENV=development
+LOG_LEVEL=info
+API_ADDRESS_DITTO=https://api.dittowords.com
```
Comment (m-evans-1) [inline .env.example:14]: 🟠 We have `docker/.env.default` and this one (`.env.example`). Ideally we should use the one in the `docker/` and update the README.  We can use [Iterable](https://github.com/lokalise/ce-connector-iterable/blob/main/README.md) as an example.

File: src/integrations/ditto/client/APIDitto.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+  private extractHeaders(apiKey: string) {
+    return {
+      Authorization: `token ${apiKey}`,
+      origin: 'lokalise',
+      'Content-Type': 'application/json',
+    }
+  }
+
+  private extractRetryConfig() {
```
Comment (kibertoad) [inline src/integrations/ditto/client/APIDitto.ts:35]: doesn't need to be a method, can easily be just a constant

File: Dockerfile (hunk 4 above)
Comment (IharKazakou) [reply Dockerfile:1]: root Dockerfile removed. docker/  Dockerfile updated

File: Dockerfile-dev (hunk 5 above)
Comment (IharKazakou) [reply Dockerfile-dev:1]: fixed

File: .env.example (hunk 6 above)
Comment (IharKazakou) [reply .env.example:14]: fixed

File: src/integrations/ditto/client/APIDitto.ts (hunk 7 above)
Comment (IharKazakou) [reply src/integrations/ditto/client/APIDitto.ts]: changed anyway

File: src/app.ts (hunk 3 above)
Comment (IharKazakou) [reply src/app.ts]: fixed

**Review by kibertoad — APPROVED** [review]

File: docker/Dockerfile [hunk 8]
```diff
@@ -27,3 +29,4 @@ RUN touch /var/log/node.log && \
 ENV PM2_WORKERS_NUMBER=max
 
 CMD /usr/local/lib/node_modules/pm2/bin/pm2-runtime start /srv/server.js -i $PM2_WORKERS_NUMBER
```
Comment (m-evans-1) [inline docker/Dockerfile:31]: 🟠 fails on PRENV

File: docker/.env.default [hunk 9]
```diff
@@ -2,17 +2,22 @@ APP_NAME=app
 CLUSTER=conn
 GIT_REPO_NAME="ce-connector-ditto"
 GIT_BASE_BRANCH=main
+HOST_APP_PORT=3000
```
[the remark this answers is not in the document]
Comment (andrewi-wd) [reply docker/.env.default:5]: New relic app name should be `live conn-ditto lok-cteng-conn-ditto.live.lokalise.cloud`. Based on connectors we already added to Newrelic

File: docker/Dockerfile (hunk 8 above)
Comment (andrewi-wd) [reply docker/Dockerfile:31]: `/srv/server.js` -> `/srv/src/server.js`

File: docker/.env.default (hunk 9 above)
[the remark this answers is not in the document]
Comment (kibertoad) [reply docker/.env.default:5]: Why 0 instead of false?

[the remark this answers is not in the document]
Comment (andrewi-wd) [reply docker/.env.default:5]: :orange_circle: We also need `NODE_ENV` variable

[the remark this answers is not in the document]
Comment (m-evans-1) [reply docker/.env.default:5]: @kibertoad mainly for consistency across the connectors. Each `.env.default` seems to be different.

[the remark this answers is not in the document]
Comment (kibertoad) [reply docker/.env.default:5]: I would standardize on the end of using `false`, as it's what we use outside of connectors. 0 for booleans is a very PHP thing, it's quite alien for JS world.

[the remark this answers is not in the document]
Comment (IharKazakou) [reply docker/.env.default:5]: @kibertoad @m-evans-1 final decision regarding 0/false?

[the remark this answers is not in the document]
Comment (kibertoad) [reply docker/.env.default:5]: I would go with false

[the remark this answers is not in the document]
Comment (IharKazakou) [reply docker/.env.default:6]: fixed

File: docker/Dockerfile (hunk 8 above)
Comment (IharKazakou) [reply docker/Dockerfile]: fixed

**Review by andrewi-wd — APPROVED** [review]

**Pull request merged**